packages/runtime/src/next/api-handler/handlers/bypass-token.ts

```diff
@@ -0,0 +1,107 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { NextRequest, NextResponse } from 'next/server'
+import { P, match } from 'ts-pattern'
+import { cookies, draftMode } from 'next/headers'
+import { parse as parseSetCookie } from 'set-cookie-parser'
+
+const SECRET_HEADER = 'x-makeswift-api-key'
+const PRERENDER_BYPASS_COOKIE = '__prerender_bypass'
+const PREVIEW_DATA_COOKIE = '__next_preview_data'
+const DRAFT_DATA_COOKIE = 'x-makeswift-draft-data'
+
+type Context = { params: { [key: string]: string | string[] } }
+
+type BypassTokenError = string
+
+type Response = { cookies: { name: string; value: string }[] }
+
+export type BypassTokenResponse = BypassTokenError | Response
+
+type BypassTokenHandlerArgs =
+  | [request: NextRequest, context: Context, params: { apiKey: string }]
+  | [req: NextApiRequest, res: NextApiResponse<BypassTokenResponse>, params: { apiKey: string }]
+
+const routeHandlerPattern = [P.instanceOf(Request), P.any, P.any] as const
+const apiRoutePattern = [P.any, P.any, P.any] as const
+
+export default async function bypassTokenHandler(
+  request: NextRequest,
+  context: Context,
+  { apiKey }: { apiKey: string },
+): Promise<NextResponse<BypassTokenResponse>>
+export default async function bypassTokenHandler(
+  req: NextApiRequest,
+  res: NextApiResponse<BypassTokenResponse>,
+  { apiKey }: { apiKey: string },
+): Promise<void>
+export default async function bypassTokenHandler(
+  ...args: BypassTokenHandlerArgs
+): Promise<NextResponse<BypassTokenResponse> | void> {
+  return match(args)
+    .with(routeHandlerPattern, args => bypassTokenRouteHandler(...args))
+    .with(apiRoutePattern, args => bypassTokenApiRouteHandler(...args))
+    .exhaustive()
+}
+
+async function bypassTokenRouteHandler(
+  request: NextRequest,
+  _context: Context,
+  { apiKey }: { apiKey: string },
+): Promise<NextResponse<BypassTokenResponse>> {
+  const secret =
+    request.nextUrl.searchParams.get(SECRET_HEADER) ?? request.headers.get(SECRET_HEADER)
```

Comment:
I think that we should have a separate constant for the search param and the header. It's a bit weird we're using the `SECRET_HEADER` constant for reading a URL search param